docs/pallas/tpu/distributed.md

```diff
@@ -0,0 +1,1453 @@
+- If two devices copy to the same destination, you may encounter non-deterministic results due to a race condition, or crashing due to  non-zero semaphore states.
+
++++ {"id": "vpGSN1Sui0Bu"}
+
+### Example: Right Permute (`lax.ppermute`)
+
+Let's dive into a very basic example. We will implement a kernel that performs a right permutation, where each device sends its slice of the data to its right neighbor.
+
+Suppose we had an array with 512 elements, which we shard into slices of size 128 across 4 devices. Each device will pass its slice to the next device, and the output will consist of the same data, but with the slices rotated by 1. This is identical to the `lax.ppermute` operation where the permutation is set to `(n, n+1)` for each device.
+
+In order to call the kernel in distributed mode, we wrap the `pallas_call` in a `shard_map` transformation. From there, we can write the kernel the same way as you would write a normal single-device Pallas kernel, except we now have access to remote DMA instructions. JAX collective primitives such as `lax.axis_index` can be used to obtain a `device_id` that can be used to compute which target devices to copy to, by referencing the same named axes names passed into `shard_map`.
+
+```{code-cell} ipython3
+---
+executionInfo:
+  elapsed: 1606
+  status: ok
+  timestamp: 1722904803566
+  user:
+    displayName: Justin Fu
+    userId: '17543197034567316452'
+  user_tz: 420
+id: YkyIKN2thZ-V
+outputId: 9b7ed142-d161-4237-fed8-cbce41adc5f0
+---
+partition = P(None, 'x')
+devices = mesh_utils.create_device_mesh((1, num_devices))
+mesh = jax.sharding.Mesh(devices, partition)
+sharding = jax.sharding.NamedSharding(mesh, partition)
+
+# Create an input array that shards the last dimension across
+# all devices.
+input_arr = jax.random.uniform(jax.random.key(0), (8, 128 * num_devices))
+input_arr = jax.device_put(input_arr, sharding)
+
+
+def right_permute_kernel(input_ref, output_ref, send_sem, recv_sem):
+  my_id = lax.axis_index('x')
+  right_neighbor = lax.rem(my_id + 1, num_devices)
+  remote_copy_op = pltpu.make_async_remote_copy(
+      src_ref=input_ref,
+      dst_ref=output_ref,
+      send_sem=send_sem,
+      recv_sem=recv_sem,
+      device_id=(0, right_neighbor),
+      device_id_type=pltpu.DeviceIdType.MESH,
+  )
+  remote_copy_op.start()
+  remote_copy_op.wait()
+
+
+out_shape = jax.ShapeDtypeStruct((8, 128), jnp.float32)
+grid_spec = pltpu.PrefetchScalarGridSpec(
+    num_scalar_prefetch=0,
+    in_specs=[
+        pl.BlockSpec(memory_space=pltpu.TPUMemorySpace.ANY),
+    ],
+    scratch_shapes=(
+        # We allocate DMA semaphores in scratch memory.
+        [pltpu.SemaphoreType.DMA]
+        * 2
+    ),
+)
+right_permute = pl.pallas_call(
+    right_permute_kernel,
+    out_shape=out_shape,
+    grid_spec=grid_spec,
+)
+# Wrap the kernel within a shard_map to call.
+pallas_result = jax.jit(
```

Comment:
Do we need these `jax.jit`s?

Reply:
No it's not necessary. Would it be better to remove?

Reply:
Not sure. I was just curious. LGTM either way!